tapa.py

```diff
@@ -0,0 +1,163 @@
+
+
+def iterate_bitmasks_for_clues(clues, num_bits):
+    """
+    Returns an iterator yielding the subset of integers in [0, 2^num_bits)
+    which, when interpreted as a bitmask, would be valid as the fill state of
+    cells surrounding a cell with the given clues.
+    """
+    bound = 1 << num_bits
+    high_bit = bound >> 1
+    mask = bound - 1
+    for bits in range(bound):
+        orig_bits = bits
+        groups = []
+        if bits == mask:
+            groups.append(num_bits)
+        elif bits != 0:
+            # Rotate the bits until the high bit is clear so we don't have to
+            # worry about bit groups wrapping around (we are guaranteed that at
+            # least one bit is clear because we just handled the bits == mask
+            # case)
+            while bits & high_bit != 0:
+                bits = (bits << 1) & mask | 1
+
+            count = 0
+            while bits > 0:
+                if bits & 1 == 1:
+                    count += 1
+                else:
+                    if count > 0:
+                        groups.append(count)
+                    count = 0
+                bits >>= 1
+            if count > 0:
+                groups.append(count)
+        if do_groups_match_clues(groups, clues):
+            yield orig_bits
+
+
+def solve_tapa(puzzle):
+    lines = [
+        [
+            [
+                char for char in cell if char != ' '
+            ] for cell in line.split('|')
+        ] for line in puzzle.strip().split('\n') if line != ''
+    ]
+
+    # Each cell in this grid will have an integer variable which is >= 0 if the
+    # cell is filled and < 0 otherwise. (The specific values within those
+    # ranges are only relevant for the one-contiguous-region constraints.)
+    g = Grid(len(lines[0]), len(lines))
+
+    def neighbor_is_filled(c):
+        return not isinstance(c, Invalid) and c.var >= 0
+
+    s = Solver()
+
+    for cell in g.cells:
+        clues = lines[cell.y][cell.x]
+        if clues:
+            # Clued cells may not be filled
+            s.add(cell.var < 0)
+
+            if clues != ['*']:
+                # Constrain the surroundings of a cell by its clues
+                surrounding_cells = get_surrounding_cells(cell)
+                num_bits = len(surrounding_cells)
+                s.add(Or([
+                    And([
+                        neighbor_is_filled(cell2) == (bits & (1 << i) != 0)
+                        for i, cell2 in enumerate(surrounding_cells)
+                    ]) for bits in iterate_bitmasks_for_clues(clues, num_bits)
+                ]))
+        else:
+            # Filled cells must form one contiguous region (part 1)
+            s.add(Or(cell.var <= 0, *(
+                And(n.var >= 0, cell.var > n.var) for n in cell.neighbors()
+            )))
+
+    # Filled cells must form one contiguous region (part 2)
+    s.add(Distinct([c.var for c in g.cells]))
+
+    # No 2×2 regions
+    for y in range(g.height - 1):
```

Comment:
This is probably easier or at least clearer what you're trying to do with 
```suggestion
    for p in g.points:
        if p.is_outside: continue
        s.add(Not(And([c.var >= 0 for e in [p.edge_left, p.edge_right] for c in e.cells()]))
```
and would be even easier if you added a `p.cells()`. (I just added `is_outside` for one of the ones I did last week.)

Reply:
Ehhh, is that clearer? I don't really think of the ‘no 2×2 regions’ restriction in terms of clusters of cells around interior points of the grid. I think my way would generalize more easily to regions of other dimensions and possibly shapes, if that's a variation that ever happens.

Reply:
It's funny that we have such totally opposite intuition. It's your code anyway; do it however you like!